File: frontend/src/engine/marketFit.js
// Market Profiles - Defined by Ideal Ranges (Mean) and Tolerance (Std Dev)
const MARKETS = {
    editorial: {
        description: "High fashion, sharp features, unconventional beauty.",
        bounty: {
            fWHR: { mean: 1.9, std: 0.15 },
            canthalTilt: { mean: 6, std: 4 },
            jawToCheekRatio: { mean: 0.9, std: 0.1 },
            midToLowerRatio: { mean: 1.0, std: 0.2 },
        }
    },
    commercial: {
        description: "Approachable, classic beauty, balanced proportions.",
        bounty: {
            fWHR: { mean: 1.75, std: 0.08 },
            canthalTilt: { mean: 2, std: 2 },
            jawToCheekRatio: { mean: 0.8, std: 0.05 },
            midToLowerRatio: { mean: 1.0, std: 0.1 },
        }
    },
    character: {
        description: "Unconventional features, distinct deviations.",
        bounty: {
            fWHR: { mean: 1.75, std: 0.3 },
            canthalTilt: { mean: 0, std: 10 },
            jawToCheekRatio: { mean: 0.75, std: 0.2 },
        }
    }
};

function gaussian(x, mean, std) {
    return (1 / (std * Math.sqrt(2 * Math.PI))) * Math.exp(-0.5 * Math.pow((x - mean) / std, 2));
}

function computeScore(metrics, profile) {
    let scoreSum = 0;
    let weightSum = 0;

    for (const [trait, stats] of Object.entries(profile.bounty)) {
        if (metrics[trait] === undefined) continue;

        const val = parseFloat(metrics[trait]);
        const p = gaussian(val, stats.mean, stats.std);
        const maxP = gaussian(stats.mean, stats.mean, stats.std);

        // Normalized score 0-1 for this trait
        const matchScore = p / maxP;

        scoreSum += matchScore;
        weightSum += 1;
    }

    return weightSum > 0 ? (scoreSum / weightSum) : 0;
}

export function calculateMarketFit(rawMetrics) {
    if (!rawMetrics) return null;
    const results = {};
    for (const [marketName, profile] of Object.entries(MARKETS)) {
        const avgFit = computeScore(rawMetrics, profile);
        results[marketName] = {
            score: Math.round(avgFit * 100),
            description: profile.description
        };
    }
    return results;
}

/**
 * Calculates a "Potential Score" simulating reduced facial puffiness.
 * Hypothesis: Lower body fat / puffiness increases Jaw Definition (JawToCheekRatio) 
 * and might slightly affect fWHR perception (more angular).
 */
export function calculatePotential(rawMetrics) {
    if (!rawMetrics) return null;

    // Simulation: Improve Jaw Definition by 10-15% towards the "Ideal" of 1.0 (squaredness)
    // or just strictly increase it if it's low.
    let currentJaw = parseFloat(rawMetrics.jawToCheekRatio);

    // Simulation: "Leaning out" increases defined jaw width relative to soft cheeks.
    const projectedJaw = currentJaw * 1.12;

    const projectedMetrics = {
        ...rawMetrics,
        jawToCheekRatio: projectedJaw
    };

    const current = calculateMarketFit(rawMetrics);
    const projected = calculateMarketFit(projectedMetrics);

    const potentialResults = {};

    for (const market in current) {
        potentialResults[market] = {
            current: current[market].score,
            potential: projected[market].score,
            gain: projected[market].score - current[market].score
        };
    }

    return potentialResults;
}

File: backend/engine/market_fit.py
import numpy as np

# Market Ideal Profiles
# Now focusing on "Golden Ratio" compliance for Commercial/Ideal
# And "Editorial" deviations.

MARKETS = {
  "ideal_golden": {
    "description": "Scientific Perfection. Adherence to Golden Ratio & Neoclassical Canons.",
    "bounty": {
      "phiRatio": {"mean": 1.618, "std": 0.12},    # Relaxed from 0.05
      "eyeSpacingRatio": {"mean": 1.0, "std": 0.15}, # Relaxed from 0.1
      "mouthNoseRatio": {"mean": 1.618, "std": 0.2},
      "fWHR": {"mean": 1.9, "std": 0.15} 
    }
  },
  "editorial": {
    "description": "High fashion. Values extreme dimorphism, long midfaces, and unique features.",
    "bounty": {
      "phiRatio": {"mean": 1.50, "std": 0.15},
      "jawToCheek": {"mean": 0.92, "std": 0.08},
      "canthalTilt": {"mean": 8.0, "std": 4.0},
      "fWHR": {"mean": 2.2, "std": 0.25},  # Tuned for Jordan Barrett (2.26)
      "midface_compactness": {"mean": 1.9, "std": 0.2}, 
      "nose_ratio": {"mean": 0.8, "std": 0.15} 
    }
  },
  "commercial": {
    "description": "Balanced, symmetrical, approachable.",
    "bounty": {
      "phiRatio": {"mean": 1.6, "std": 0.08},
      "eyeSpacingRatio": {"mean": 1.0, "std": 0.08},
      "jawToCheek": {"mean": 0.82, "std": 0.05},
      "midface_compactness": {"mean": 2.1, "std": 0.1} # Prefers compact
    }
  }
}

def gaussian(x, mean, std):
    return np.exp(-0.5 * ((x - mean) / std) ** 2)

def calculate_market_fit(metrics: dict) -> dict:
    results = {}
    for market, profile in MARKETS.items():
        total_score = 0
        count = 0
        
        for trait, stats in profile["bounty"].items():
            if trait not in metrics: continue
            val = metrics[trait]
            score = gaussian(val, stats["mean"], stats["std"])
            total_score += score
            count += 1
            
        avg_score = (total_score / count) if count > 0 else 0
        results[market] = {
            "score": int(avg_score * 100),
            "description": profile["description"],
            "traits": profile["bounty"]
        }
    return results

def calculate_potential(metrics: dict) -> dict:
    if not metrics: return None
    
    # Simulate Leanness / Optimization
    # 1. Jaw Definition improves (+15%)
    # 2. Phi Ratio might shift slightly due to less cheek fat (bizygoma reduces -> Phi increases?)
    #    - If cheek fat loss reduces Bizygoma width, FaceHeight/Bizygoma INCREASES.
    #    - This usually moves people closer to 1.618 if they are wide (low ratio).
    
    projected = metrics.copy()
    
    # Jaw improves
    projected["jawToCheek"] = min(0.98, projected.get("jawToCheek", 0.75) * 1.15)
    
    # Cheek fat loss (Bizygoma reduction simulation)
    # If Bizygoma reduces by 2%, Phi Ratio increases by ~2%
    curr_phi = projected.get("phiRatio", 1.5)
    projected["phiRatio"] = curr_phi * 1.02
    
    # Recalculate
    curr_fit = calculate_market_fit(metrics)
    proj_fit = calculate_market_fit(projected)
    
    potential_analysis = {}
    for m in curr_fit:
        potential_analysis[m] = {
            "pre_score": curr_fit[m]["score"],
            "post_score": proj_fit[m]["score"],
            "gain": proj_fit[m]["score"] - curr_fit[m]["score"],
            "note": "Simulated low body fat & reduced puffiness."
        }
        
    return potential_analysis
